### PR TITLE
Ferinko/checkpoint gone

### PR DESCRIPTION
**Pull request by Ferinko**
This PR primarily implements an optimisation that trades raw number of modules for decreased complexity of modules by splitting on preconditions of functions.

### CODE REVIEW

File: src/Horus/CairoSemantics.hs [hunk 1]
```diff
@@ -417,7 +417,12 @@ mkCallConstraints pc nextPc fp optimisingF f = do
   -- Considering we have already pushed the stackFrame by now, we need to make sure that either
   -- the function is inlinable and we'll encounter a 'ret', or we need to pop right away
   -- once we encode semantics of the function.
-  if Just f == ((^. _3) <$> optimisingF)
+
+  -- Determine whether the current function matches the function being optimised exactly -
+  -- this necessitates comparing execution traces.
+  executionContext <- getStackTraceDescr
+  optimisedFExecutionContext <- getStackTraceDescr' ((^. _1) <$> optimisingF)
+  if isJust optimisingF && executionContext == optimisedFExecutionContext
```
Comment (Ferinko) [inline src/Horus/CairoSemantics.hs:425]: Mmm, how would I write this better in Haskell?

In Lean 4, I could do `<- getStackTraceDescr == <- getStackTraceDescr' ((^. _1) <$> optimisingF)`.
Of course I can do `getStackTraceDescr >>= \exec` but then I'm already polluting the context with these `exec`s, just like in this do-expanded version.

File: src/Horus/CairoSemantics.hs [hunk 2]
```diff
@@ -402,30 +468,31 @@ mkApConstraints apEnd insts = do
...
       getApIncrement fp lInst >>= \case
-        Just apIncrement -> let res = assert (ap1 + apIncrement .== ap2) in res
+        Just apIncrement -> assert (ap1 + apIncrement .== ap2)
         Nothing | not isNewStackframe -> assert (ap1 .< ap2)
         Nothing -> pure ()
   lastAp <- getAp lastPc
   when (isRet lastInst) pop
   fp <- getFp
   getApIncrement fp lastLInst >>= \case
-    Just lastApIncrement -> let res = assert (lastAp + lastApIncrement .== apEnd) in res
```
Comment (Julek) [inline src/Horus/CairoSemantics.hs:412]: How did this weird pattern end up getting several times into the code base? 0_0

**Review by Julek — CHANGES_REQUESTED** [review]
Minor comments

File: src/Horus/CairoSemantics.hs (hunk 1 above)
Comment (Ferinko) [reply src/Horus/CairoSemantics.hs:425]: Well, the entire subsequent scope is polluted by `executionContext` and `optimisedExecutionContext`, I don't really want to have the ability to refer to these, I don't think.

File: src/Horus/CFGBuild.hs [hunk 3]
```diff
@@ -101,6 +149,20 @@ getInvariant name = liftF' (GetInvariant name id)
...
+
+getVerts :: Label -> CFGBuildL [Vertex]
+getVerts l = liftF' (GetVerts l id)
+
+-- It is enforced that for any one PC, one can add at most a single salient vertex
+getSalientVertex :: Label -> CFGBuildL Vertex
+getSalientVertex l = do
+  verts <- filter (not . isOptimising) <$> getVerts l
+  -- This can be at most one, so len <> 1 implies there are no vertices
+  unless (length verts == 1) . throw $ "No vertex with label: " <> tShow l
```
[the remark this answers is not in the document]
Comment (Ferinko) [reply src/Horus/CFGBuild.hs:163]: This is embarrassing :).

File: src/Horus/CairoSemantics.hs (hunk 2 above)
Comment (Ferinko) [reply src/Horus/CairoSemantics.hs:412]: I use this pattern to do: `let res = huge_expr in trace huge_expr res` and then I removed the trace and failed to fix the code :).

**Review by langfield — CHANGES_REQUESTED** [review]
In general, it would probably be good policy to add descriptive docstrings similar to those you'd see on Hackage to all nontrivial functions which are nontrivially edited in PRs. Especially in `CairoSemantics.hs`, which is probably the most confusing module.

There is an argument to be made that code doesn't need comments if it's self-documenting to its audience. I won't debate whether or not that's true in general, it probably holds under certain conditions. But in this case, your audience is literally just me (and Julian and Ilia to a lesser extent), and it's definitely not self-documenting.

Also, please rename `storageRemoval` to something a bit longer that describes what's actually going on. Something like `substituteStorageReadsWithConditionals` or something?

There are some queries and such pushed to intermediate commits. Another thing that would make working on the project nicer: treating every commit that will get merged to `master` like a release. Every commit should build, pass CI, not have any debugging artifacts, be readable.

Also, several of these commits have sweeping changes to multiple things. These commit messages... Okay of course it is much better for each revision to have a title than to have nothing, but I don't really think that's sufficient here. The point of the commit messages is so that if someone who is not you wants to understand what's going on in a diff, they don't have to get on a call with you to get English-language context. I think you could write a couple thousand words on what's going on just in the first commit of this PR and not be done.

Look at how the `git` maintainers use commit messages:

```console
commit 068fa54c0034be0b953d798628b06815f9cfaff0
Author: Derrick Stolee <derrickstolee@github.com>
Date:   Tue Jul 19 15:26:06 2022 +0000

    midx: reduce memory pressure while writing bitmaps

    We noticed that some 'git multi-pack-index write --bitmap' processes
    were running with very high memory. It turns out that a lot of this
    memory is required to store a list of every object in the written
    multi-pack-index, with a second copy that has additional information
    used for the bitmap writing logic.

    Using 'valgrind --tool=massif' before this change, the following chart
    shows how memory load increased and was maintained throughout the
    process:

        GB
    4.102^                                                             ::
         |              @  @::@@::@@::::::::@::::::@@:#:::::::::::::@@:: :
         |         :::::@@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |      :::: :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |    :::: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |    : :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |    : :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |   :: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |   :: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |   :: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |   :: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |   :: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |   :: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         |   :: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         | @ :: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         | @ :: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         | @::: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         | @::: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         | @::: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
         | @::: :: : :: @@:@: @ ::@ ::: ::::@: ::: @@:#:::::: :: : :@ :: :
       0 +--------------------------------------------------------------->

    It turns out that the 'struct write_midx_context' data is persisting
    through the life of the process, including the 'entries' array. This
    array is used last inside find_commits_for_midx_bitmap() within
    write_midx_bitmap(). If we free (and nullify) the array at that point,
    we can free a decent chunk of memory before the bitmap logic adds more
    to the memory footprint.

    Here is the massif memory load chart after this change:

        GB
    3.111^      #
         |      #                              :::::::::::@::::::::::::::@
         |      #        ::::::::::::::::::::::::: : :: : @:: ::::: :: ::@
         |     @#  :::::::::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |     @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |  :::@#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |  :: @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |  :: @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
         |  :: @#::: ::: :::::: :::: :: : :::::::: : :: : @:: ::::: :: ::@
       0 +--------------------------------------------------------------->

    The previous change introduced a refactoring of write_midx_bitmap() to
    make it more clear how much of the 'struct write_midx_context' instance
    is needed at different parts of the process. In addition, the following
    defensive programming measures were put in place:

     1. Using FREE_AND_NULL() we will at least get a segfault from reading a
        NULL pointer instead of a use-after-free.

     2. 'entries_nr' is also set to zero to make any loop that would iterate
        over the entries be trivial.

     3. Add significant comments in write_midx_internal() to add warnings
        for future authors who might accidentally add references to this
        cleared memory.

    Signed-off-by: Derrick Stolee <derrickstolee@github.com>
    Signed-off-by: Junio C Hamano <gitster@pobox.com>
```

They have a graph in there.

**Pull request closed without merging**